arrows/vxl/color_commonality_filter.cxx

```diff
@@ -0,0 +1,600 @@
+  template < class InputType, class OutputType >
+  typename std::enable_if< std::is_integral< OutputType >::value >::type
+  filter_color_image( vil_image_view< InputType > const& input,
+                      vil_image_view< OutputType >& output,
+                      std::vector< unsigned >& histogram,
+                      color_commonality_filter_settings const& options );
+
+  template < class InputType, class OutputType >
+  void
+  perform_filtering( vil_image_view< InputType > const& input,
+                     vil_image_view< OutputType >& output,
+                     color_commonality_filter_settings const& options );
+
+  template < typename pix_t >
+  kwiver::vital::image_container_sptr
+  compute_commonality( vil_image_view< pix_t >& input );
+
+  color_commonality_filter* p;
+
+  color_commonality_filter_settings settings;
+
+  unsigned color_resolution{ 512 };
+  unsigned color_resolution_per_chan{ 8 };
+  unsigned intensity_resolution{ 16 };
+
+  std::vector< unsigned > color_histogram;
+  std::vector< unsigned > intensity_histogram;
+};
+
+// ----------------------------------------------------------------------------
+template < class InputType, class OutputType >
+typename std::enable_if< std::is_integral< OutputType >::value >::type
+
+color_commonality_filter::priv
+::filter_color_image( vil_image_view< InputType > const& input,
+                      vil_image_view< OutputType >& output,
+                      std::vector< unsigned >& histogram,
+                      color_commonality_filter_settings const& options )
+{
+  if( input.ni() != output.ni() || input.nj() != output.nj() )
+  {
+    LOG_ERROR( p->logger(),
+               "Input and output images must be the same dimensions." );
+    return;
+  }
+  if( !is_power_of_two( options.resolution_per_channel ) )
+  {
+    LOG_ERROR( p->logger(),
+               "The resolution per channel must be a power of two." );
+    return;
+  }
+
+  if( input.ni() == 0 || input.nj() == 0 )
+  {
+    return;
+  }
+
+  // Configure output scaling settings based on the output type and user
+  // settings
+  InputType const input_type_max = std::numeric_limits< InputType >::max();
+  OutputType const output_type_max = std::numeric_limits< OutputType >::max();
+  auto const histogram_threshold = static_cast< unsigned >( output_type_max );
+  unsigned histogram_scale_factor = options.output_scale_factor;
+
+  // Use type default options if no scale factor specified
+  if( histogram_scale_factor == 0 )
+  {
+    histogram_scale_factor = histogram_threshold;
+  }
+
+  // Populate histogram steps for each channel of the hist
+  std::vector< std::ptrdiff_t > hist_steps{ input.nplanes() };
+  hist_steps[ 0 ] = 1;
+
+  for( unsigned p = 1; p < input.nplanes(); ++p )
+  {
+    hist_steps[ p ] = hist_steps[ p - 1 ] * options.resolution_per_channel;
+  }
+
+  // Fill in histogram of the input image
+  unsigned const bitshift =
+    integer_log2( input_type_max + 1 ) -
```

Comment:
Uh... I think this is still broken for... actually, I'm not sure it works, ever, but I think it may just need:
```suggestion
    integer_log2( input_type_max ) + 1 -
```
Also, have you (at least mentally) verified this produces the correct result for `bitshift`? I haven't gone through the mental exercise of running this code in my head to verify that it isn't off-by-one; maybe you've done that already?

OTOH, I guess the tests are passing? I'd feel better knowing this has been micro-validated, though. If I'm wrong (always possible!), I'd at least feel better knowing someone is confident this is working.

Reply:
I validated that this produces the correct bitshift, and the tests should fail violently if it's off. Since `integer_log2` truncates, and `input_type_max` will always be `2^n - 1`, the two suggested methods will behave the same as far as I can tell. I just thought this way was cleaner.

Reply:
Maybe I'm missing something, or maybe you're getting lucky with integer promotion (and not testing with 32- or 64-bit integers), but won't `input_type_max + 1` be `input_type_min` in at least some cases? Yes, it will be the same _if no overflow occurs_...

Reply:
Yup, you're right. I'll update it. 

The tests are only run on byte images, and AFAIK burnout has never really been pushed past that, so maybe that's something to consider in the future.

Reply:
Yeah, and I think you luck out because `char{255} + 1` produces `int{256}` due to integer promotion rules. I'm a little surprised actually there isn't a compiler warning when a constant expression overflows, but there are probably also instances of people doing that deliberately.

Reply:
Okay, with the unit tests and the most recent update, I am reasonably confident this is working. Thanks!